File: kattis/bracketmatching.rb
gets; inp = gets.chomp
stack = []
opening_brackets = ['(', '[', '{']
closing_brackets = [')', ']', '}']
bracket_pairs = { ')' => '(', ']' => '[', '}' => '{' }
inp.each_char do |char|
    if opening_brackets.include?(char)
        stack.push(char)
    elsif closing_brackets.include?(char)
        if stack.empty? || stack.pop != bracket_pairs[char]
            puts 'Invalid'
            exit
        end
    end
end
puts stack.empty? ? "Valid" : "Invalid"

File: kattis/bracketmatching.py
input(); inp = input()
stack = []
for char in inp:
    if (char == ")" or char == "]" or char == "}") and not len(stack):
        print("Invalid")
        exit(0)
    if char == ")" and stack[-1] == "(" or char == "]" and stack[-1] == "[" or char == "}" and stack[-1] == "{":
        stack.pop()
    elif char == ")" or char == "]" or char == "}":
        print("Invalid")
        exit(0)
    else:
        stack.append(char)
print("Valid" if not len(stack) else "Invalid")